### PR TITLE
model scaffolding and local model inferencing

### PR DESCRIPTION
**Pull request by y-f-u**
This PR adds inferencing for onnx supported model.
- introduce modelruntime - tract
- modelformat - onnx
- modelsource - local (spice.ai to be implemented)
- and a predictions api

### CODE REVIEW

File: crates/runtime/src/modelruntime/tract.rs [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+    pub path: String,
+}
+
+type Plan = SimplePlan<TypedFact, Box<dyn TypedOp>, Graph<TypedFact, Box<dyn TypedOp>>>;
+pub struct Model {
+    model: Plan,
+}
+
+impl Model {
+    fn run_inference(
```
Comment (Jeadie) [inline crates/runtime/src/modelruntime/tract.rs]: nit: We should just put this straight into `fn run(&self, input: Vec<RecordBatch>) -> super::Result<RecordBatch> {
`

File: crates/runtime/src/model.rs [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+            }
+            _ => UnknownDataSourceSnafu {
+                model_source: source,
+            }
+            .fail()?,
+        }
+    }
+
+    pub async fn run(&self, df: Arc<DataFusion>) -> Result<RecordBatch> {
+        let sql = format!("select * from datafusion.public.{}", self.datasets[0]);
```
Comment (Jeadie) [inline crates/runtime/src/model.rs]: If we set `order by 'ts' asc`, it should make the inference simpler (currently we have to do this in the inference function.

File: crates/runtime/src/model.rs [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+    pub fn load(model: &spicepod::component::model::Model) -> Result<Self> {
+        let source = model.source();
+        let source = source.as_str();
+
+        let mut params = std::collections::HashMap::new();
+        params.insert("name".to_string(), model.name.to_string());
+        params.insert("from".to_string(), model.from.to_string());
+
+        match source {
+            "local" => {
```
Comment (Jeadie) [inline crates/runtime/src/model.rs]: don't we need to match on `spiceai`? Probably a good candidate for an enum so compiler will naturally catch it.

Comment (y-f-u) [reply crates/runtime/src/model.rs]: will do this right after; when implementing the spiceai source.

File: crates/runtime/src/http/v1.rs [hunk 4]
```diff
@@ -35,3 +35,65 @@ pub(crate) mod datasets {
...
+        response::{IntoResponse, Response},
+        Extension, Json,
+    };
+    use serde::Serialize;
+    use std::sync::Arc;
+    use std::time::Instant;
+    use tract_core::tract_data::itertools::Itertools;
+
+    #[derive(Serialize)]
+    pub struct JsonResponse {
```
Comment (phillipleblanc) [inline crates/runtime/src/http/v1.rs]:
```suggestion
    pub struct ForecastResponse {
```

File: crates/runtime/src/model.rs (hunk 3 above)
Comment (phillipleblanc) [inline crates/runtime/src/model.rs]:
```suggestion
            "localhost" => {
```

Match the dataset format

File: crates/runtime/src/modelruntime.rs [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+    TractError { source: tract_core::anyhow::Error },
+
+    ArrowError { source: arrow::error::ArrowError },
+
+    ShapeError { source: ndarray::ShapeError },
+}
+pub type Result<T, E = Error> = std::result::Result<T, E>;
+
+pub trait Runnable: Send + Sync {
+    fn run(&self, input: Vec<RecordBatch>) -> Result<RecordBatch>;
```
Comment (phillipleblanc) [inline crates/runtime/src/modelruntime.rs]: For traits, I've come to the conclusion that it doesn't make sense to define concrete error types - as you then need to leak implementation details about which error can occur in the trait definition. It's better to just return `std::result::Result<RecordBatch, Box<dyx std:error::Error>>`

Comment (y-f-u) [reply crates/runtime/src/modelruntime.rs]: that makes more sense. I prefer what you mentioned, working on the change

Comment (phillipleblanc) [reply crates/runtime/src/modelruntime.rs]: To be clear, the actual implementations should still use Snafu - but the trait itself doesn't need it.

File: crates/runtime/src/modelsource/local.rs [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+use std::collections::HashMap;
+use std::sync::Arc;
+
+impl ModelSource for Local {
+    fn pull(&self, params: Arc<Option<HashMap<String, String>>>) -> super::Result<String> {
+        let name = params
+            .as_ref()
+            .as_ref()
+            .and_then(|p| p.get("name"))
+            .map(std::string::ToString::to_string);
```
Comment (phillipleblanc) [inline crates/runtime/src/modelsource/local.rs]: Add `use std::string` and then change this to `ToString::to_string`

File: crates/runtime/src/modelsource/local.rs [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+
+        let Some(name) = name else {
+            return Err(super::Error::UnableToCreateModelPath {
+                source: std::io::Error::new(
+                    std::io::ErrorKind::Other,
+                    "Unable to create model path",
+                ),
+            });
```
Comment (phillipleblanc) [inline crates/runtime/src/modelsource/local.rs]: This shouldn't be an io error

File: crates/runtime/src/modelsource/local.rs [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+        };
+
+        let _ = super::ensure_model_path(name.as_str());
```
Comment (phillipleblanc) [inline crates/runtime/src/modelsource/local.rs]: Check the error?

File: crates/runtime/src/modelsource/local.rs [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+
+        let path = params
+            .as_ref()
+            .as_ref()
+            .and_then(|p| p.get("from"))
+            .map(std::string::ToString::to_string);
```
Comment (phillipleblanc) [inline crates/runtime/src/modelsource/local.rs]:
```suggestion
            .map(ToString::to_string);
```

File: crates/runtime/src/modelsource/local.rs [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+
+        let Some(path) = path else {
+            return Err(super::Error::UnableToCreateModelPath {
+                source: std::io::Error::new(
+                    std::io::ErrorKind::Other,
+                    "Unable to create model path",
+                ),
```
Comment (phillipleblanc) [inline crates/runtime/src/modelsource/local.rs]: Also not an IO error

File: crates/runtime/src/modelsource.rs [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+        source: Box<dyn std::error::Error + Send + Sync>,
+    },
+
+    #[snafu(display("Unable to find home directory"))]
+    UnableToFindHomeDir {},
+
+    #[snafu(display("Unable to create model path"))]
+    UnableToCreateModelPath { source: std::io::Error },
+
+    #[snafu(display("Unable to create model path"))]
```
Comment (phillipleblanc) [inline crates/runtime/src/modelsource.rs]:
```suggestion
    #[snafu(display("Unable to load the configuration"))]
```

File: crates/runtime/src/modelsource/local.rs [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+            .as_ref()
+            .and_then(|p| p.get("name"))
+            .map(ToString::to_string);
+
+        let Some(name) = name else {
+            return Err(super::UnableToLoadConfigSnafu {}.build());
+        };
+
+        // it is not copying local model into .spice folder
+        let _ = super::ensure_model_path(name.as_str())?;
```
Comment (phillipleblanc) [inline crates/runtime/src/modelsource/local.rs:21]: This should use a Snafu error

**Review by phillipleblanc — APPROVED** [review]

File: crates/runtime/src/modelsource/local.rs (hunk 12 above)
Comment (y-f-u) [reply crates/runtime/src/modelsource/local.rs:21]: it already returns a snafu error.

**Pull request merged**